### PR TITLE
OCPBUGS-5523: Catalog, fatal error: concurrent map read and map write

### PR DESCRIPTION
**Pull request by dtfranz**
Protected subscriptionSyncCounters access to prevent concurrent map writes

**Description of the change:**

The metrics pkg uses a shared map var which is not protected from concurrent read/write access. This PR adds a mutex to lock and unlock before and after each map access instance. 

**Motivation for the change:**

OCPBUGS-5523

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -199,12 +200,37 @@ var (
...
+
+func (s *subscriptionSyncCounter) setValues(key string, val subscriptionSyncLabelValues) {
+	s.countersLock.Lock()
+	defer s.countersLock.Unlock()
+	s.counters[key] = val
+}
+
+func (s *subscriptionSyncCounter) readValues(key string) (subscriptionSyncLabelValues, bool) {
+	s.countersLock.Lock()
+	defer s.countersLock.Unlock()
```
Comment (joelanford) [inline pkg/metrics/metrics.go:229]:
```suggestion
	s.countersLock.RLock()
	defer s.countersLock.RUnlock()
```

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -199,12 +200,37 @@ var (
...
+	subscriptionSyncCounters = newSubscriptionSyncCounter()
 )
 
+// subscriptionSyncCounter keeps a record of the Prometheus counters emitted by
+// Subscription objects. The key of a record is the Subscription name, while the value
+// is struct containing label values used in the counter. Read and Write access are
+// protected by mutex.
+type subscriptionSyncCounter struct {
+	counters     map[string]subscriptionSyncLabelValues
+	countersLock sync.Mutex
```
[the remark this answers is not in the document]
Comment (dtfranz) [reply pkg/metrics/metrics.go]: Updated.

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (dtfranz) [reply pkg/metrics/metrics.go]: Updated.

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -199,12 +200,37 @@ var (
...
+	s.countersLock.Lock()
+	defer s.countersLock.Unlock()
+	s.counters[key] = val
+}
+
+func (s *subscriptionSyncCounter) readValues(key string) (subscriptionSyncLabelValues, bool) {
+	s.countersLock.RLock()
+	defer s.countersLock.RUnlock()
+	val, ok := s.counters[key]
+	return val, ok
```
Comment (awgreene) [inline pkg/metrics/metrics.go:231]:
```suggestion
	return s.counters[key]
```

**Review by awgreene — APPROVED** [review]
Seems reasonable to me, could we squash the commits? It'll make it easier for downstreaming.

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -199,12 +200,37 @@ var (
...
+	subscriptionSyncCounters = newSubscriptionSyncCounter()
 )
 
+// subscriptionSyncCounter keeps a record of the Prometheus counters emitted by
+// Subscription objects. The key of a record is the Subscription name, while the value
+// is struct containing label values used in the counter. Read and Write access are
+// protected by mutex.
+type subscriptionSyncCounter struct {
+	counters     map[string]subscriptionSyncLabelValues
+	countersLock sync.RWMutex
```
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/metrics/metrics.go:212]: /hold
Should this be a sync.Mutex? I believe that hte sync.RWMutex requires that only a single routine is classified as the writer, and I believe that controllers can be configured to run multiple routines in parallel. I can't recall if we configured OLM to reconcile subscriptions in parallel offhand, but I'll check tomorrow.

[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/metrics/metrics.go:212]: This is incorrect, only a single writer can HOLD the lock at a given time.
/unhold

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (awgreene) [reply pkg/metrics/metrics.go:231]: Wow, I would have thought that this behaved the same way as it does for [functions](https://go.dev/play/p/J6HLJ0ZZQH_6), but for some reason [the bool isn't returned](https://go.dev/play/p/JSvaVQCXiKX) doesn't work when trying to use the result of a map lookup to satisfy

Comment (awgreene) [reply pkg/metrics/metrics.go:231]: I suspect this is because you don't have to receive the value as shown [here](https://go.dev/play/p/gXjkM1_zVPs).

Comment (awgreene) [reply pkg/metrics/metrics.go:231]: Courtesy of @benluddy:

[From the Go Docs](https://go.dev/ref/spec#Index_expressions):
> An index expression on a map a of type map[K]V used in an [assignment statement](https://go.dev/ref/spec#Assignment_statements) or initialization of the special form
> v, ok = a[x]
> v, ok := a[x]
> var v, ok = a[x]
> yields an additional untyped boolean value.

**Review by tmshort — APPROVED** [review]
/lgtm

**Comment by awgreene** [comment]
@dtfranz this needs a rebase.

**Review by tmshort — APPROVED** [review]
/lgtm

**Pull request merged**